rider-fsharp/src/main/java/com/jetbrains/rider/plugins/fsharp/services/fsi/FsiSandboxInfoUpdater.kt

```diff
@@ -0,0 +1,152 @@
+package com.jetbrains.rider.plugins.fsharp.services.fsi
+
+import com.intellij.execution.process.ProcessAdapter
+import com.intellij.execution.process.ProcessEvent
+import com.intellij.execution.process.ProcessOutputTypes
+import com.intellij.openapi.editor.event.EditorFactoryEvent
+import com.intellij.openapi.editor.event.EditorFactoryListener
+import com.intellij.openapi.editor.ex.EditorEx
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Key
+import com.jetbrains.rdclient.editors.FrontendTextControlHost
+import com.jetbrains.rdclient.editors.sandboxes.SandboxManager
+import com.jetbrains.rd.platform.util.application
+import com.jetbrains.rd.util.lifetime.Lifetime
+import com.jetbrains.rd.util.lifetime.LifetimeDefinition
+import com.jetbrains.rdclient.lang.toRdLanguageOrThrow
+import com.jetbrains.rdclient.util.idea.fromOffset
+import com.jetbrains.rider.editors.RiderTextControlHost
+import com.jetbrains.rider.ideaInterop.fileTypes.fsharp.FSharpScriptLanguage
+import com.jetbrains.rider.model.*
+import com.jetbrains.rider.projectView.solution
+import org.jetbrains.concurrency.AsyncPromise
+
+class FsiSandboxInfoUpdater(
+        private val project: Project, private val consoleEditor: EditorEx, private val history: CommandHistory) {
+
+    private val rdFsiTools = project.solution.rdFSharpModel.fSharpInteractiveHost.fsiTools;
+
+    private val lockObject = Object()
+
+    private var processLifetimeDefinition : LifetimeDefinition? = null
+    private var processLifetime : Lifetime? = null
+
+    val fsiProcessOutputListener = FsiSandboxInfoUpdaterProcessOutputListener(this)
+
+    private var verifiedCommandNumber = 0
+    private val preparedCommands = mutableListOf<String>()
+    private var lastPreparedCommandIndex: Int? = null
+    private val correctCommandNumbers = mutableListOf<Int>()
+
+    private fun updateSandboxInfo() {
+        application.invokeLater {
+            val sandboxManager = SandboxManager.getInstance()
+            if (sandboxManager.getSandboxInfo(consoleEditor) == null) return@invokeLater
+
+            var startUnpreparedCommandIndex = 0
+            val endUnpreparedCommandIndex = correctCommandNumbers.size
+            if (lastPreparedCommandIndex != null) {
+                startUnpreparedCommandIndex = lastPreparedCommandIndex!! + 1
```

Comment:
Can `lastPreparedCommandIndex.Size` be used instead of mutating a boxed nullable int?